presto-router/src/main/java/com/facebook/presto/router/scheduler/WeightedRoundRobinScheduler.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.router.scheduler;
+
+import com.facebook.airlift.log.Logger;
+
+import java.net.URI;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Collections.nCopies;
+
+// As the weighted round-robin scheduler, similar to the normal round-robin method,
+// keeps the selected index as a state for scheduling, the candidates and weights
+// shall not be modified after they are assigned.
+// This design indicates that the weighted round-robin scheduler can only be used
+// when the candidates and weights are always consistent. Otherwise, users need to
+// create a new scheduler object to pick up the change.
+public class WeightedRoundRobinScheduler
+        implements Scheduler
+{
+    private List<URI> candidates;
+    private HashMap<URI, Integer> weights;
+    private List<URI> serverList;
+
+    private static Integer serverIndex = 0;
+    private static final Logger log = Logger.get(WeightedRoundRobinScheduler.class);
+
+    @Override
+    public Optional<URI> getDestination(String user)
+    {
+        checkArgument(candidates.size() == weights.size());
+
+        if (serverList == null) {
+            this.generateServerList();
+        }
+
+        try {
+            synchronized (serverIndex) {
+                if (serverIndex >= serverList.size()) {
+                    serverIndex = 0;
+                }
+                return Optional.of(serverList.get(serverIndex++));
+            }
+        }
+        catch (IllegalArgumentException e) {
+            log.warn(e, "Error getting destination for user " + user);
+            return Optional.empty();
+        }
+    }
+
+    public void setCandidates(List<URI> candidates)
+    {
+        // Only keeps the first given `candidates` due to maintaining the
```

Comment:
in case the candidates list is immutable,  how about to move it to the constructor and set it to be 'final'?

Reply:
That's a nice point! 
The candidate list for this specific scheduler is immutable, but for others, it's not. As the `Scheduler` interface needs to be compatible with other child classes (so child class objects could be created under a uniform interface), I don't make the candidate list immutable in this class. If you have any suggestions for overcoming the design challenge, feel free to let me know!